### ISSUE TITLE
Parse FinanceEvents

### ISSUE THREAD

**Issue by mbolder** [python-amazon-mws/python-amazon-mws#170]
Hello, I am stuck trying to parse the results from list_financial_events report.
I have got two problems, a.) how do I check if an event is empty 
and b.) how do iterate over the dicts in order to create one set of information per order?

So far I worked on this, but I get the error "list indices must be integers or slices, not str"

`
def print_dict(v, prefix=''):
    if isinstance(v, dict):
        for k, v2 in v.items():
            p2 = "{}['{}']".format(prefix, k)
            print_dict(v2, p2)
    elif isinstance(v, list):
        for i, v2 in enumerate(v):
            p2 = "{}[{}]".format(prefix, i)
            print_dict(v2, p2)
    else:
        print('{} = {}'.format(prefix, repr(v)))

def getAmaFinanceEvents(groupID):
    finance_api = mws.Finances(access_key, secret_key, merchant_id, region="DE")
    events = finance_api.list_financial_events(financial_event_group_id=groupID)
    events = events.parsed
    print_dict(events)
    Subscriptionfee = events.FinancialEvents.ServiceFeeEventList.ServiceFeeEvent.FeeList.FeeComponent.FeeAmount.CurrencyAmount
    i=0
    for event in enumerate(events.FinancialEvents.ShipmentEventList.ShipmentEvent):
        AmazonOrderID = events['FinancialEvents']['ShipmentEventList']['ShipmentEvent'][i]['AmazonOrderId']['value']
        Principal =     events['FinancialEvents']['ShipmentEventList']['ShipmentEvent'][i]['ShipmentItemList']['ShipmentItem']['ItemChargeList']['ChargeComponent'][0]['ChargeAmount']['CurrencyAmount']['value']
        ShippingCharge =events['FinancialEvents']['ShipmentEventList']['ShipmentEvent'][i]['ShipmentItemList']['ShipmentItem']['ItemChargeList']['ChargeComponent'][4]['ChargeAmount']['CurrencyAmount']['value']
        i=i+1
        print (AmazonOrderID,Principal,ShippingCharge)
`

Any thoughts or comments would greatly appreciated.

**Comment by GriceTurrble** [comment]
Based on the [example response](http://docs.developer.amazonservices.com/en_US/finances/Finances_ListFinancialEvents.html) in the docs for this request, I assume the error you're getting comes from the `ShipmentItem` key, which may be a list if there is more than one item.

Additionally, it's odd that you're iterating `for event in enumerate(events.FinancialEvents.ShipmentEventList.ShipmentEvent)`, but never using either `event` or the index you receive from the `enumerate` generator. Instead, you're switching to dict keys instead of attributes, beginning again from the `events` root object; and you're using a pattern with `i=0` and `i=i+1`, which is not even necessary when using [`enumerate`](https://docs.python.org/3.8/library/functions.html#enumerate).

I would suggest the following, which I hope demonstrates how to work with the `ObjectDict` nodes more cleanly:

```python
for event in events.FinancialEvents.ShipmentEventList.ShipmentEvent:
    amazon_order_id = event.AmazonOrderId
    for shipment_item in event.ShipmentItemList.ShipmentItem:
        for charge_component in shipment_item.ItemChargeList.ChargeComponent:
            charge_type = charge_component.ChargeType
            # check here `if charge_type == "Principal"` to get the principal? Or other types as needed?
            charge_currency_amount = charge_component.ChargeAmount.CurrencyAmount
            charge_currency_code = charge_component.ChargeAmount.CurrencyCode
```

Each node is iterable by default, which is the best way to approach those nodes which may be a list of more than one item. In each of the `for` loops above, the right-most node (`ShipmentItem`, `ChargeComponent`, etc) is one that may be some list of multiple entries in the XML response. This is safer than manually accessing the dict keys and list indices, since we don't always know which type that node will be (single instances become dicts, where using a list index would cause an exception; and vice versa, multiple instances become lists, where using a dict key would cause an exception).

As seen above, assigning one of those nodes to some new variable just makes it easier to access the elements further beneath it, without having to repeat node names as often.

Finally, I wouldn't assume that Amazon returns a list of ChargeComponents in the same order every time, so using magic numbers for list indices `0` and `4` can get you into trouble. Better to iterate that node and check the `ChargeType` to get the information you want.

**Comment by mbolder** [comment]
Thank you for the fast reply !! 
I must be blind, now I understand the approach but my code leads to  
order_item_id = shipment_item.OrderItemId
AttributeError: 'str' object has no attribute 'OrderItemId'

`
for shipment_event in events.FinancialEvents.ShipmentEventList.ShipmentEvent:
        amazon_order_id = shipment_event.AmazonOrderId
        for shipment_item in shipment_event.ShipmentItemList.ShipmentItem:
            order_item_id = shipment_item.OrderItemId
            for charge_component in shipment_item.ItemChargeList.ChargeComponent:
                charge_type = charge_component.ChargeType
                if charge_type=="Principal":
                    principal = charge_component.ChargeAmount.CurrencyAmount
            #for fee_component in shipment_item.ItemFeeList.FeeComponent:
                #fee_type = fee_component.FeeType
`

**Comment by GriceTurrble** [comment]
Visual Studio Code is my preference.

Perhaps my reading of the example response in the docs doesn't match the output you're getting. I would just recommend playing with the response object in your interactive interpreter to find the data you require.

In particular, `ShipmentItem` according to docs should be a node with other child elements below it. If you see a different bit of data in your response, that may just be Amazon not following their documented method. :man_shrugging:

**Comment by mbolder** [comment]
It always stops at the second list, here is an example ShipmentEvent I get from Amazon :

```
for ShipmentEvent in events.FinancialEvents.ShipmentEventList.ShipmentEvent:
                print (ShipmentEvent)

```
{'value': '\n          ', 'ShipmentItemList': {'value': '\n            ', 'ShipmentItem': {'value': '\n              ', 'ItemChargeList': {'value': '\n                ', 'ChargeComponent': [{'value': '\n                  ', 'ChargeType': {'value': 'Principal'}, 'ChargeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '34.9'}, 'CurrencyCode': {'value': 'EUR'}}}, {'value': '\n                  ', 'ChargeType': {'value': 'Tax'}, 'ChargeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}}, {'value': '\n                  ', 'ChargeType': {'value': 'GiftWrap'}, 'ChargeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}}, {'value': '\n                  ', 'ChargeType': {'value': 'GiftWrapTax'}, 'ChargeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}}, {'value': '\n                  ', 'ChargeType': {'value': 'ShippingCharge'}, 'ChargeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '7.0'}, 'CurrencyCode': {'value': 'EUR'}}}, {'value': '\n                  ', 'ChargeType': {'value': 'ShippingTax'}, 'ChargeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}}]}, 'ItemFeeList': {'value': '\n                ', 'FeeComponent': [{'value': '\n                  ', 'FeeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '-5.24'}, 'CurrencyCode': {'value': 'EUR'}}, 'FeeType': {'value': 'Commission'}}, {'value': '\n                  ', 'FeeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}, 'FeeType': {'value': 'FixedClosingFee'}}, {'value': '\n                  ', 'FeeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}, 'FeeType': {'value': 'GiftwrapCommission'}}, {'value': '\n                  ', 'FeeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '-1.05'}, 'CurrencyCode': {'value': 'EUR'}}, 'FeeType': {'value': 'ShippingHB'}}, {'value': '\n                  ', 'FeeAmount': {'value': '\n                    ', 'CurrencyAmount': {'value': '0.0'}, 'CurrencyCode': {'value': 'EUR'}}, 'FeeType': {'value': 'VariableClosingFee'}}]}, 'OrderItemId': {'value': 'xxxxxxxxxxxxxxx'}, 'QuantityShipped': {'value': '1'}, 'SellerSKU': {'value': 'xx-xxxx-xxxx'}}}, 'AmazonOrderId': {'value': 'xxx-xxxxxx-xxxxxxx'}, 'PostedDate': {'value': '2020-03-29T16:13:02.990Z'}, 'MarketplaceName': {'value': 'Amazon.de'}}

Maybe you could have a look, in case you have minute.

**Comment by mbolder** [comment]
I think I got a bit closer to finding the problem, the ShipmentItemList is not always a list. When ever there was only one item sold it, is not a list, only another node. If there are more than one item in the shipment it is a list. How would I differentiate here ?

**Comment by Bobspadger** [comment]
Sorry I'm a bit late to this party - but here is a little helper function I use for exactly this purpost.

```python
def listify(amazon_object):
    """
    Test if returned amazon item is a list, if not, return it as a list.

    :param amazon_object:

    :return:
    """
    if isinstance(amazon_object, list):  # pylint: disable=no-else-return
        return amazon_object
    else:
        return [amazon_object]
```

**Comment by mbolder** [comment]
Yes, that is a better approach, instead of having to handle two sets of events at that point.
I tested both against our data and was surprised that my approach was faster, but it seems because we sell mostly one item at a time, the list operation takes its time. Anyway it makes the code much easier to understand and maintain.  
Now that the bookkeeping is sorted any tips with the order reports ?

**Comment by Bobspadger** [comment]
Yep - there are some issues around here where we have discussed it. I'll try to dig some out in the next day or so.

I have a cron job that runs and requests / picks up reports and actions them.

Its worth reading the MWS documentation on how this flow should work.

**Comment by mbolder** [comment]
Oh, one more question that came up during the work on the FinancialEvents, is there a way to check if an event is empty ? So far is just catch the error e.g. :

```
for Event in events.FinancialEvents:
        if Event=="ServiceFeeEventList":
            try:
                Subscriptionfee = float(events.FinancialEvents.ServiceFeeEventList.ServiceFeeEvent.FeeList.FeeComponent.FeeAmount.CurrencyAmount)
            except KeyError:
                Subscriptionfee = 0
```

But that feels a bit clumsy......

**Comment by Bobspadger** [comment]
You should be able to use the `.get()` method

https://docs.python.org/3/library/stdtypes.html?highlight=get#dict.get

this will return a default value if the key does not exist, so no need to capture `KeyError`

Hope that helps!

Alex

**Comment by Bobspadger** [comment]
Just as an update to this which may help:

because it is a nested dictionary structure, you need to chain the `.get()` method together, and return an empty dictionary for the items further up the chain to the key you wish to get.

Here is an example from a similar situation where I go looking for products in amazon:


```python
    for lpl in LowestOfferListing.get('LowestOfferListing'):
        amz = lpl.get('Qualifiers', {}).get('FulfillmentChannel', {}).get('value')
```
this would equate to the dot notation structure of

`lpl.Qualifiers.FulfillmentChannel.get('value)`

and makes sure that if any of the intermediate dictionaries do not exists / don't have the key, you don't get an unexpected exception :)
